### PR TITLE
Provided fix for issue 6365

### LINKED ISSUE #6365: X2 Handoff - UE ping/pong between sites ultimately causing ICS failures
### Your Environment

- **Version:** 1.5
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** MME
- **Deployment Environment:** AWS (orc8r), local docker-compose (orc8r), bare-metal (AGW), 

### Describe the Issue
* UE Does a handoff from src_eNB (100.96.228.117) to target_enb (100.96.228.116). PathSwitchRequest is successfully acknowledged (timestamp  19:32:16 in MME log); eNB/MME S1AP IDs (2/35)
* UE transitions to old eNB and sends SR (6.3 seconds after PathSwitchAck; packet 177 in pcap) and uses S1AP IDs (2/49). Confirmed this by matching the mTMSI shown in SR with mme log.
* target_enb also sends ICS request using IDs (2/35) (Pkt 179). 
* MME releases context in target eNB 4 seconds later. 
* eNB resets S1AP IDs 2/49. 
* UE sends SR again (same mTMSI as above) in Pkt 196 using eNB_S1AP_ID 50.
* MME attempts to setup IC on src enb (100.96.228.117) using new S1AP IDs (2/50). 
* eNB sends ICS failure due to "encryption-and-or-integrity-protection-algorithms-not-supported"
* Context Release procedure commences; MME re-sends Context Release command in packet 205 for previously released context. S1AP IDs 2/49 were reset in packets 193 and 194. 

Comparing ICS Requests between packet 198 (failed one) and a good one (packet 12), it seems that `encryptionAlgorithms` and `integrityProtectionAlgorithms` are all set to `0`. This might be the issue? 

image
image

**To Reproduce**
* Do back and forth ping-pongs between eNBs connected to the same AGW. 

**Expected behavior**
* ICS failure needs to be mitigated
* Context release commands for already acknowledged S1-AP Reset messages should be handled properly? not sure on the phrasing here but I think MME shouldn't be releasing this context after acknowledging reset request previously. 

[Archive.zip](https://github.com/magma/magma/files/6368074/Archive.zip)

PCAP and MME log attached.

### PR DESCRIPTION
**Pull request by rsarwad**
"[lte][agw]: Provided fix for issue 6365

## Summary
After X2 handover, for forth coming Service Request sent from UE was resulting into ICS failure. Because 
1. Security capabilities in s1ap is defined as uint16_t and security capabilities defined with emm_context is of uint8_t; 
2. Due to this mismatch while handling path switch request, security capabilities were getting copied from path switch request message. While copying only LSB byte was getting copied. From S1ap message;  path switch request perspective, LSB bits are reserved fields and so always zero was getting copied to security capability of emm context.

## Test Plan
Tested with s1ap sanity test suite

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c [hunk 1]
```diff
@@ -3630,9 +3630,12 @@ void mme_app_handle_path_switch_request(
...
   ue_context_p->e_utran_cgi       = path_switch_req_p->ecgi;
   ue_network_capability.eea =
-      path_switch_req_p->encryption_algorithm_capabilities;
+      ((path_switch_req_p->encryption_algorithm_capabilities >> 9) |
+       ((0x80 & ue_context_p->emm_context._ue_network_capability.eea)));
   ue_network_capability.eia =
-      path_switch_req_p->integrity_algorithm_capabilities;
+      ((path_switch_req_p->integrity_algorithm_capabilities >> 9) |
+       ((0x80 & ue_context_p->emm_context._ue_network_capability.eia)));
+
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c:3647]: Can you comment on why it is updated in this particular manner for code readability and preventing someone to make accidental changes in this logic?

**Review by ulaskozat — APPROVED** [review]
LGTM with a nit comment.

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c [hunk 2]
```diff
@@ -3629,10 +3629,22 @@ void mme_app_handle_path_switch_request(
   }
   ue_context_p->sctp_assoc_id_key = path_switch_req_p->sctp_assoc_id;
   ue_context_p->e_utran_cgi       = path_switch_req_p->ecgi;
+
+  /* Security capabilities IE within s1ap message, Path Switch Request is of
+   * 16 bit info, encryption algorithms starts with 128-EEA1
+   * second bit set to 128-EEA2,
+   * third bit set to 128-EEA3, other bits are reserved.
+   * Where as, eee within emm context is of 8 bit info and starts from EEA0
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c]: eee -> eea

**Comment by rdefosse** [comment]
@rsarwad 

The OAI pipeline run `https://jenkins-oai.eurecom.fr/job/MAGMA-MME-production/3433/` is a false positive.

A GitHub service was down when your CI run occurred. It meant that the pipeline was in bypass mode.

This will need to be re-triggered.

Sorry for the inconvenience
Raphael

**Comment by rdefosse** [comment]
@rsarwad 

Your last run is perfect! Thx

`https://jenkins-oai.eurecom.fr/job/MAGMA-MME-production/3440/`

Raphael

**Review by themarwhal — APPROVED** [review]

**Comment by rsarwad** [comment]
The PR is ready2merge

**Pull request merged**